// @lwc/compiler/dist/commonjs/compiler/compiler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("@lwc/errors");
const bundler_1 = require("../bundler/bundler");
const options_1 = require("./options");
const index_1 = require("../index");
var template_compiler_1 = require("@lwc/template-compiler");
exports.templateCompiler = template_compiler_1.default;
async function compile(options) {
    options_1.validateOptions(options);
    const normalizedOptions = options_1.normalizeOptions(options);
    let result;
    const diagnostics = [];
    const { diagnostics: bundleDiagnostics, code, map, metadata, } = await bundler_1.bundle(normalizedOptions);
    diagnostics.push(...bundleDiagnostics);
    if (!hasError(diagnostics)) {
        result = {
            code,
            map,
            metadata,
            outputConfig: normalizedOptions.outputConfig,
        };
    }
    return {
        version: index_1.version,
        success: !hasError(diagnostics),
        diagnostics,
        result
    };
}
exports.compile = compile;
function hasError(diagnostics) {
    return diagnostics.some(d => {
        return d.level === errors_1.DiagnosticLevel.Error || d.level === errors_1.DiagnosticLevel.Fatal;
    });
}
