// @lwc/engine/lib/shared/static-node-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const language_1 = require("./language");
const fields_1 = require("./fields");
const Items = fields_1.createFieldName('items');
class StaticNodeList extends NodeList {
    item(index) {
        return this[index];
    }
    get length() {
        return fields_1.getInternalField(this, Items).length;
    }
    // Iterator protocol
    forEach(cb, thisArg) {
        language_1.forEach.call(fields_1.getInternalField(this, Items), cb, thisArg);
    }
    entries() {
        return language_1.ArrayMap.call(fields_1.getInternalField(this, Items), (v, i) => [i, v]);
    }
    keys() {
        return language_1.ArrayMap.call(fields_1.getInternalField(this, Items), (v, i) => i);
    }
    values() {
        return fields_1.getInternalField(this, Items);
    }
    [Symbol.iterator]() {
        let nextIndex = 0;
        return {
            next: () => {
                const items = fields_1.getInternalField(this, Items);
                return nextIndex < items.length ?
                    {
                        value: items[nextIndex++], done: false
                    } : {
                    done: true
                };
            }
        };
    }
}
function createStaticNodeList(items) {
    const nodeList = language_1.create(StaticNodeList.prototype, {
        [Items]: {
            value: items,
        }
    });
    // setting static indexes
    language_1.forEach.call(items, (item, index) => {
        language_1.defineProperty(nodeList, index, {
            value: item,
            enumerable: true,
            configurable: true,
        });
    });
    return nodeList;
}
exports.createStaticNodeList = createStaticNodeList;
